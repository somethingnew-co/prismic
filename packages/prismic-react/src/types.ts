import React from 'react'
import { Elements, HTMLSerializer } from 'prismic-reactjs'
import { LinkResolver } from '@stnew/prismic-types'

export type SliceMap = Map<string, React.ReactType | Promise<any>>

interface Slices {
  [key: string]: React.ReactType | Promise<any>
}

export interface Resolvers {
  linkResolver: LinkResolver
  hrefResolver?: LinkResolver
}

export interface Provider extends Resolvers {
  slices?: Slices
  htmlSerializer?: Hash
}

export interface Context extends Resolvers {
  sliceMap?: SliceMap
  htmlSerializer?: HTMLSerializer<React.ReactNode>
}

// HTMLSerializer types
export interface PropsObject {
  [prop: string]: any
}
export interface Propagator {
  (element?: any): PropsObject
}

export type Hash = {
  [E in Elements]?: [React.ReactType, (PropsObject | Propagator)?]
}
