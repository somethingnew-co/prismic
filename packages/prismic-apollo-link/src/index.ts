export { default as prismicApolloLink } from './apolloLink'
