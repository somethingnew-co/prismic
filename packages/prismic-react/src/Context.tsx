import { createContext } from 'react'
import { Context } from './types'

const prismicContext: Context = {
  slices: {},
  linkResolver: () => '/',
  hrefResolver: () => '/',
  rootResolver: '',
}

const PrismicContext = createContext(prismicContext)

export default PrismicContext
