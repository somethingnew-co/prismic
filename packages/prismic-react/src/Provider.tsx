import React from 'react'
import PrismicContext from './Context'
import { Provider, Context } from './types'

function PrismicProvider({ slices, linkResolver, hrefResolver, children }: Provider): JSX.Element {

  const sliceMap = new Map(Object.entries(slices))

  const value: Context = {
    sliceMap,
    linkResolver,
    hrefResolver,
  }

  return <PrismicContext.Provider value={value}>{children}</PrismicContext.Provider>
}

export default PrismicProvider
