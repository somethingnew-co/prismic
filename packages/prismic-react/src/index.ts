export { default as PrismicContext } from './Context'
export { default as PrismicProvider } from './Provider'
export { default as usePrismic } from './usePrismic'
export { default as useHtmlSerializer } from './useHtmlSerializer'
export { SliceZone } from './SliceZone'
export { default as htmlSerializerThunk } from './utils/serializerThunk'
export * from '@stnew/prismic'
