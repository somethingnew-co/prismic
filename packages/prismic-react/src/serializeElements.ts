import React from 'react'
import { Elements, HTMLSerializer } from 'prismic-reactjs'
import { HTMLSerializerElementMap } from './types'

function serializeElements(map: HTMLSerializerElementMap): HTMLSerializer<React.ReactNode> {
  return function (type, element, content, children, key): React.ReactNode {

    const node = map[type as Elements]

    if (node) {
      const [component, props] = node
      let propagator = props || {}

      if (typeof props === 'function') {
        propagator = props(element)
      }

      const propsWithKey = { ...propagator, key }

      // <img /> elements don't have children
      if (type === Elements.image) {
        return React.createElement(component, propsWithKey)
      }

      return React.createElement(component, propsWithKey, children)
    }
  }
}

export default serializeElements
