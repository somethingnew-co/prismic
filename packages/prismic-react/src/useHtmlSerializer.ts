import React from 'react'
import { HTMLSerializerElementMap } from './types'
import { HTMLSerializer } from 'prismic-reactjs'
import serializerThunk from './utils/serializerThunk'

function useHtmlSerializer(hash: HTMLSerializerElementMap): HTMLSerializer<React.ReactNode> {
  return serializerThunk(hash)
}

export default useHtmlSerializer
