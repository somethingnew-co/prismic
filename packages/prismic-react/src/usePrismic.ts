import { useContext } from 'react'
import PrismicContext from './Context'
import { Context } from './types'

function usePrismic(): Context {
  const value = useContext(PrismicContext)
  return value
}

export default usePrismic
