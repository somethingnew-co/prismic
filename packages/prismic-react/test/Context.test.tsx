import React from 'react'
import { PrismicContext } from '../src'
import { render } from '@testing-library/react'
import { linkResolver, hrefResolver } from '../../../test/resolvers'
import { docs, slices } from '../../../test/data'
import TestSlice from '../../../test/slices/TestSlice'
import { PrismicDoc } from '@stnew/prismic-types'

test('Context', () => {
  const testDoc = docs.find(doc => doc.type === 'blog_post')

  const { container } = render(
    <PrismicContext.Provider value={{
      sliceMap: new Map(Object.entries({
        slice: TestSlice,
      })),
      linkResolver,
      hrefResolver,
    }}>
      <PrismicContext.Consumer>
        {(value) => {
          const SliceComponent = value.sliceMap?.get('slice') as React.ReactType

          if (!SliceComponent) return null

          return (
            <div>
              <SliceComponent {...slices[0]}/>
              <div>{value.linkResolver(testDoc as PrismicDoc)}</div>
              <div>{value.hrefResolver && value.hrefResolver(testDoc as PrismicDoc)}</div>
            </div>
          )}}
      </PrismicContext.Consumer>
    </PrismicContext.Provider>,
  )

  expect(container).toMatchSnapshot()

})
