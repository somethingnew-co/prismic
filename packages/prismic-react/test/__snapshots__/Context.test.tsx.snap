// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Context 1`] = `
<div>
  <div>
    <div>
      <h1>
        Hello World
      </h1>
    </div>
    <div>
      /blog/hello-world
    </div>
    <div>
      /blog/[post]
    </div>
  </div>
</div>
`;
